internal/pkg/v2/infrastructure/redis/event.go

```diff
@@ -0,0 +1,73 @@
+//
+// Copyright (C) 2020 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package redis
+
+import (
+	"encoding/json"
+
+	"github.com/edgexfoundry/edgex-go/internal/pkg/common"
+
+	model "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+
+	"github.com/gomodule/redigo/redis"
+	"github.com/google/uuid"
+)
+
+const EventsCollection = "v2:event"
+
+// ************************** DB HELPER FUNCTIONS ***************************
+func addEvent(conn redis.Conn, e model.Event) (addedEvent model.Event, err error) {
+	if e.Created == 0 {
+		e.Created = common.MakeTimestamp()
+	}
+
+	if e.Id == "" {
+		e.Id = uuid.New().String()
+	}
+
+	event := model.Event{
+		CorrelationId: e.CorrelationId,
+		Checksum:      e.Checksum,
+		Id:            e.Id,
+		Pushed:        e.Pushed,
+		DeviceName:    e.DeviceName,
+		Created:       e.Created,
+		Modified:      e.Modified,
+		Origin:        e.Origin,
+	}
+
+	m, err := json.Marshal(event)
+
+	_ = conn.Send("MULTI")
+	// use the SET command to save event as blob
+	_ = conn.Send("SET", EventsCollection+":"+e.Id, m)
+	_ = conn.Send("ZADD", EventsCollection, 0, e.Id)
+	_ = conn.Send("ZADD", EventsCollection+":created", e.Created, e.Id)
+	_ = conn.Send("ZADD", EventsCollection+":pushed", e.Pushed, e.Id)
+	_ = conn.Send("ZADD", EventsCollection+":deviceName:"+e.DeviceName, e.Created, e.Id)
+	if e.Checksum != "" {
+		_ = conn.Send("ZADD", EventsCollection+":checksum:"+e.Checksum, 0, e.Id)
+	}
+	// add reading ids as sorted set under each event id
+	// sort by the order provided by device service
+	rids := make([]interface{}, len(e.Readings)*2+1)
+	rids[0] = EventsCollection + ":readings:" + e.Id
+	for i, r := range e.Readings {
+		readingId, err := addReading(conn, r)
```

Comment:
The `deviceName` of readings are assigned in the `ToReadingModel` function in https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/v2/dtos/reading.go#L51-L57.

As for the `Created` time of reading which is a timestamp saved in DB is defined in the `checkReadingValue` function
in `internal/pkg/v2/infrastructure/redis/reading.go`
```
func checkReadingValue(b *model.BaseReading) (err error) {
	if b.Created == 0 {
		b.Created = common.MakeTimestamp()
	}
```
https://github.com/edgexfoundry/edgex-go/pull/2610/files#diff-bec3ad46361da938c3c8bb25f916e4d6R58-R61

Reply:
I am good on handling of `Created` THX!
Since we don't have any validation on the on the Reading DTO, we are not guaranteed that it is set. 
https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/v2/dtos/reading.go#L25
I am ok with addressing this with more robust validation on the AddEventRequest & Reading DTOs 
If you agree, please submit issue on go-core-contracts for `robust validation on the AddEventRequest & Reading DTOs`. THX.

Reply:
Thanks for the suggestion. PRs https://github.com/edgexfoundry/edgex-go/pull/2653 and https://github.com/edgexfoundry/go-mod-core-contracts/pull/264 are opened for the suggested validation enhancements.